drivers/stepper/adi/tmc5041_stepper_controller.c

```diff
@@ -0,0 +1,608 @@
+				controller_config->int_pin_spec.port->name);
+			return -ENODEV;
+		}
+
+		status = gpio_pin_configure_dt(&controller_config->int_pin_spec, GPIO_INPUT);
+		if (status) {
+			LOG_ERR("Failed to configure INT pin GPIO %d", status);
+			return status;
+		}
+
+		gpio_init_callback(&controller_data->int_cb, int_pin_callback_handler,
+				   BIT(controller_config->int_pin_spec.pin));
+
+		status = gpio_add_callback(controller_config->int_pin_spec.port,
+					   &controller_data->int_cb);
+		if (status < 0) {
+			LOG_ERR("Could not add INT pin GPIO callback (%d)", status);
+			return status;
+		}
+
+		status = gpio_pin_interrupt_configure_dt(&controller_config->int_pin_spec,
+							 GPIO_INT_EDGE_RISING);
+		if (status < 0) {
+			LOG_ERR("failed to configure INT interrupt (err %d)", status);
+			return -EINVAL;
+		}
+		LOG_DBG("INT pin callback set");
+		uint32_t read_value;
+		(void)tmc5041_read(dev, TMC5041_RAMPSTAT(0), &read_value);
+		(void)tmc5041_read(dev, TMC5041_RAMPSTAT(1), &read_value);
+		k_work_init(&controller_data->int_pin_callback_work, callback_work_handler);
+	}
+
+	return status;
+}
+#endif
+
+static int tmc5041_stepper_enable(const struct device *dev, const bool enable)
+{
+	LOG_DBG("Stepper motor controller %s %s", dev->name, enable ? "enabled" : "disabled");
+	const struct tmc5041_stepper_config *config = dev->config;
+	uint32_t reg_value = 0;
+	(void)tmc5041_read(config->controller, TMC5041_CHOPCONF(config->index), &reg_value);
+	if (enable) {
+		reg_value |= TMC5041_CHOPCONF_DRV_ENABLE_MASK;
+	} else {
+		reg_value &= ~TMC5041_CHOPCONF_DRV_ENABLE_MASK;
+	}
+	(void)tmc5041_write(config->controller, TMC5041_CHOPCONF(config->index), reg_value);
+	return 0;
+}
+
+static int tmc5041_stepper_is_moving(const struct device *dev, bool *is_moving)
+{
+	const struct tmc5041_stepper_config *config = dev->config;
+	uint32_t reg_value = 0;
+	(void)tmc5041_read(config->controller, TMC5041_VACTUAL(config->index), &reg_value);
+	*is_moving = reg_value ? true : false;
+	LOG_DBG("Stepper motor controller %s is moving: %d", dev->name, *is_moving);
+
+	return 0;
+}
+
+static int tmc5041_stepper_move(const struct device *dev, const int32_t steps,
+				struct k_poll_signal *async)
+{
+	const struct tmc5041_stepper_config *config = dev->config;
+	struct tmc5041_stepper_data *data = dev->data;
+
+	(void)set_async_signal(dev, async);
+
+#ifdef CONFIG_STEPPER_ADI_TMC_STALLGUARD
+	stallguard_enable(dev, false);
+#endif
+	int32_t position = 0;
+	(void)stepper_get_actual_position(dev, &position);
+	int32_t target_position = position + steps;
+	(void)tmc5041_write(config->controller, TMC5041_RAMPMODE(config->index),
+			    TMC5041_RAMPMODE_POSITIONING_MODE);
+	LOG_DBG("Stepper motor controller %s moved to %d by steps: %d", dev->name, target_position,
+		steps);
+	(void)tmc5041_write(config->controller, TMC5041_XTARGET(config->index), target_position);
+#ifdef CONFIG_STEPPER_ADI_TMC_STALLGUARD
+	(void)k_work_reschedule(&data->stallguard_dwork,
```

Comment:
The stall guard activation delay is more speed dependent.
Am not sure if having a fixed delay is a good idea here.

Reply:
yep you are correct, i am introducing 
`.stallguard_threshold_velocity = DT_PROP_OR(child, stallguard_threshold_velocity, 0),`

Reply:
so the stallguard work handler would retry activation again and again like this. Normally in the drive stall should be detected and appropriately calibrated, if that is not happening then the device is anyhow not correctly calibrated.

	"err = stallguard_enable(data->stepper, true);
	if (err == -EAGAIN) {
		LOG_ERR("Trying again to enable stallguard");
		struct tmc5041_stepper_config *stepper_config = data->stepper->config;
		k_work_reschedule(dwork, K_MSEC(stepper_config->stallguard_delay_in_ms));
	}
	if( err == -EIO) {
		LOG_ERR("Failed to enable stallguard becaoause of I/O error");
		return;
	}"